quickwit-serve/src/lib.rs

```diff
@@ -21,77 +21,104 @@ mod args;
+use quickwit_search::{start_searcher_service, SearchService};
 use quickwit_storage::quickwit_storage_uri_resolver;
-use tracing::{debug, info};
 
 pub use crate::args::ServeArgs;
-use crate::cluster_api::GrpcClusterAdapter;
 pub use crate::counters::COUNTERS;
-use crate::grpc::start_grpc_service;
 #[cfg(test)]
 use crate::rest::recover_fn;
-use crate::rest::start_rest_service;
-use crate::search_api::GrpcSearchAdapter;
 
-/// Starts a search node, aka a `searcher`.
-pub async fn run_searcher(
-    quickwit_config: QuickwitConfig,
-    metastore: Arc<dyn Metastore>,
-) -> anyhow::Result<()> {
-    SEARCHER_CONFIG_INSTANCE
-        .set(quickwit_config.searcher_config.clone())
-        .expect("could not set searcher config in global once cell");
-    let cluster = Arc::new(Cluster::new(
-        quickwit_config.node_id.clone(),
-        quickwit_config.gossip_socket_addr()?,
-    )?);
-    for seed_socket_addr in quickwit_config.seed_socket_addrs()? {
-        // If the peer seed address is specified,
-        // it joins the cluster in which that node participates.
-        debug!(peer_seed_addr = %seed_socket_addr, "Add peer seed node.");
-        cluster.add_peer_node(seed_socket_addr).await;
+#[derive(Debug, PartialEq, Eq, Copy, Hash, Clone)]
+pub enum QService {
+    Indexer,
+    Searcher,
+}
+
+impl TryFrom<&str> for QService {
+    type Error = anyhow::Error;
+
+    fn try_from(service_str: &str) -> Result<Self, Self::Error> {
+        match service_str {
+            "indexer" => Ok(QService::Indexer),
+            "searcher" => Ok(QService::Searcher),
+            _ => {
+                bail!("Service `{service_str}` unknown");
+            }
+        }
     }
-    let storage_uri_resolver = quickwit_storage_uri_resolver().clone();
-    let client_pool = SearchClientPool::create_and_keep_updated(cluster.clone()).await;
-    let cluster_client = ClusterClient::new(client_pool.clone());
-    let search_service = Arc::new(SearchServiceImpl::new(
-        metastore,
-        storage_uri_resolver,
-        cluster_client,
-        client_pool,
-    ));
-
-    let cluster_service = Arc::new(ClusterServiceImpl::new(cluster.clone()));
-
-    let grpc_addr = quickwit_config.grpc_socket_addr()?;
-    let grpc_search_service =
-        GrpcSearchAdapter::from(search_service.clone() as Arc<dyn SearchService>);
-    let grpc_cluster_service = GrpcClusterAdapter::from(cluster_service.clone());
-    let grpc_server = start_grpc_service(grpc_addr, grpc_search_service, grpc_cluster_service);
-
-    let rest_socket_addr = quickwit_config.rest_socket_addr()?;
-    let rest_server = start_rest_service(rest_socket_addr, search_service, cluster_service);
-    info!(
-        "Searcher ready to accept requests at http://{}/",
-        rest_socket_addr
-    );
+}
+struct QuickwitServices {
+    pub cluster_service: Arc<dyn ClusterService>,
+    pub search_service: Option<Arc<dyn SearchService>>,
+    pub indexer_service: Option<Arc<IndexingServerClient>>,
```

Comment:
we could rename `IndexingServerClient` in `IndexingService`

Reply:
Absolutely. I want to do that soon.

Reply:
I opened #1206 accordingly.